Add comments and docs to task schedules (#2265)
pulp/pulpcore: docs/plugins/reference/task-scheduling.rst

.. _task-scheduling:

Task Scheduling
---------------

.. warning: This feature is in tech-preview and only accessible by direct manipulation of

>    ``TaskSchedule`` objects.

Reviewer: do you think it's worth mentioning that this will not be available through the API and mainly is targeted for the plugin writers usage?

> Pulp supports scheduling of tasks.

Reviewer 2: Can the docs style be switched to not line wrap? I believe that's the documented style in the style part of the docs.

Author: Oh, this is draft, i'll be wrapping them the moment we agreed on every single word. Not wanting to rewrap them over and over.

> Scheduled tasks will be dispatched shortly after their ``next_dispatch`` time, and be rescheduled one ``dispatch_interval`` after that, if the latter is set.

Reviewer 2: I guess I didn't realize `dispatch_interval` was optional. I figured we didn't support one-off tasks, but anyways this reads well.

Author: It was to easy to also get one off schedules that way to just leave it out.

By specifying the ``dispatch_interval`` as ``time_delta(days=1)`` you can expect the task dispatch to stabily happen at same time every day.

> Unless the last task of the same schedule did enter a final state, no new task will be dipatched.

Reviewer 3: typo "dipatched"

Reviewer 2: ```suggestion
If the last task of the same schedule did enter a final state, no new task will be dispatched.
```

Author: Apart from the typo the sentence was correct. Not entering a final state will block this very schedule from recurring.

Author: ```suggestion
Until the last task of the same schedule enters a final state, a new task will not be dispatched.
```

> Unless the last task of the same schedule did enter a final state, no new task will be dipatched.

Reviewer: Unless the last task enters... no new task will be...
or
If the last task does not enter...no new task will be..

conditionals :sob:

> Scheduling is done by the worker processes, therefore schedules will be missed if all workers are offline.

Reviewer 2: ```suggestion
Scheduling is done by the worker processes, therefore scheduled task dispatching will be missed if all workers are offline.
```

Author: BTW, at the same time there will be no worker to handle the tasks anyway.

> After an outage window, overdue schedules will dispatch at most one task, but down to timing, they my be rescheduled shortly thereafter.

Reviewer: s/my be/may be

Author: ```suggestion
After an outage window, overdue schedules will dispatch at most one task, but down to timing, they may be rescheduled shortly thereafter.
```

> The task schedule API at ``/pulp/api/v3/task-schedules/`` is provided to monitor the states of tasks schedules.

Reviewer 3: +1 good explanation

> The task schedule API at ``/pulp/api/v3/task-schedules/`` is provided to monitor the states of tasks schedules.

Reviewer 2: ```suggestion
The task schedule API at ``/pulp/api/v3/task-schedules/`` is provided to read the tasks schedules.
```

Author: BTW, at the same time there will be no worker to handle the tasks anyway.

> The task schedule API at ``/pulp/api/v3/task-schedules/`` is provided to monitor the states of tasks schedules.

Reviewer: I keep forgetting what 'guidelines' we follow when we choose name for new endpoints. Specifically whether to use underscore or hyphens. It seems like 'access policies is written through underscore and now we use hyphens :see_no_evil:

Author: Oh snap! I can see both. With a clear winner of underscore. But still it's probably too late to "fix" the use at `task-groups`, `signing-services` and `import-check`...

Reviewer: yeah..agreed.
